addons/delivery/models/product_template.py

```diff
@@ -11,3 +12,11 @@ class ProductTemplate(models.Model):
         string="HS Code",
         help="Standardized code for international shipping and goods declaration. At the moment, only used for the FedEx shipping provider.",
     )
+
+    def write(self, values):
```

Comment:
maybe add sudo if user have not access to this method.
maybe check for all delivery.carrier
maybe put this in a @api.contraints
maybe check for all products (1 sql query)

```python
   def write(self, vals):
        if vals.get('sale_ok') is False or vals.get('purchase_ok') is False:
                carriers = self.env['delivery.carrier'].sudo().\
                    search([('product_id', 'in', self.product_variant_ids.ids)])
                if carriers:
                    raise UserError(_('This product is set as delivery product of thoses shipping methods:\n%s', '\n'.join(carriers.mapped('display_name')))
        return super().write(vals)
```

Reply:
Thanks for your input, I did indeed take your advice into consideration.
Kr